packages/services/test/src/utils.ts

```diff
@@ -454,3 +679,119 @@ class TerminalTester extends SocketTester {
   private _onMessage: (msg: TerminalSession.IMessage) => void = null;
 }
 
+
+/**
+ * Test a single emission from a signal.
+ *
+ * @param signal - The signal we are listening to.
+ * @param find - An optional function to determine which emission to test,
+ * defaulting to the first emission.
+ * @param test - An optional function which contains the tests for the emission.
+ * @param value - An optional value that the promise resolves to if it is successful.
+ *
+ * @returns a promise that rejects if the function throws an error (e.g., if an
+ * expect test doesn't pass), and resolves otherwise.
+ *
+ * #### Notes
+ * The first emission for which the find function returns true will be tested in
+ * the test function. If the find function is not given, the first signal
+ * emission will be tested.
+ *
+ * You can test to see if any signal comes which matches a criteria by just
+ * giving a find function. You can test the very first signal by just
+ * giving a test function. And you can test the first signal matching the
+ * find criteria by giving both.
+ */
+export
+async function testEmission<T, U, V>(signal: ISignal<T, U>, options: {
+  find?: (a: T, b: U) => boolean
+  test?: (a: T, b: U) => void,
+  value?: V
+}): Promise<V> {
+  const done = new PromiseDelegate<V>();
+  const object = {};
+  signal.connect((sender: T, args: U) => {
+    if (!options.find || options.find(sender, args)) {
+      try {
+        Signal.disconnectReceiver(object);
+        if (options.test) {
+          options.test(sender, args);
+        }
+      } catch (e) {
+        done.reject(e);
+        // throw e;
+      }
+      done.resolve(options.value || undefined);
+    }
+  }, object);
+  return done.promise;
```

Comment:
Why does this function need to be asychronous?

Reply:
I added a comment. Basically, the thing that effected the signal may be asynchronous, so this allows that thing to happen.